v3/src/components/graph/components/graph.tsx

```diff
@@ -31,105 +32,133 @@ interface IProps {
+          .attr('x', xScale?.range()[0] + margin.left)
+          .attr('y', 0)
+          .attr('width', layout.plotWidth)
+          .attr('height', layout.plotHeight)
+      }
+    }, [layout.plotHeight, layout.plotWidth, margin.left, xScale])
 
-  useEffect(function setupPlotArea() {
-    if (xScale && xScale?.range().length > 0) {
-      select(plotAreaSVGRef.current)
-        .attr('x', xScale?.range()[0] + margin.left)
-        .attr('y', 0)
-        .attr('width', layout.plotWidth)
-        .attr('height', layout.plotHeight)
+    const toast = useToast()
+    const handleDropAttribute = (place: AxisPlace, attrId: string) => {
+      const attrName = dataset?.attrFromID(attrId)?.name
+      toast({
+        position: "top-right",
+        title: "Attribute dropped",
+        description: `The attribute ${attrName || attrId} was dropped on the ${place} axis!`,
+        status: "success"
+      })
+      graphModel.setAttributeID(place, attrId)
     }
-  }, [layout.plotHeight, layout.plotWidth, margin.left, xScale])
+    // respond to assignment of new attribute ID
+    useEffect(function handleNewAttributeID() {
+      const disposer = graphModel && onAction(graphModel, action => {
+        if (action.name === 'setAttributeID') {
+          const place: AxisPlace = action.args?.[0],
+            attrID = action.args?.[1]
+          enableAnimation.current = true
+          graphController.handleAttributeAssignment(place, attrID)
+          casesRef.current = graphModel.cases
+        }
+      }, true)
+      return () => disposer?.()
+    }, [graphController, dataset, layout, xAxisModel, yAxisModel, enableAnimation, graphModel])
 
-  const toast = useToast()
-  const handleDropAttribute = (place: AxisPlace, attrId: string) => {
-    const attrName = dataset?.attrFromID(attrId)?.name
-    toast({
-      position: "top-right",
-      title: "Attribute dropped",
-      description: `The attribute ${attrName || attrId} was dropped on the ${place} axis!`,
-      status: "success"
-    })
-    graphModel.setAttributeID(place, attrId)
-  }
-  // respond to assignment of new attribute ID
-  useEffect(function handleNewAttributeID() {
-    const disposer = graphModel && onAction(graphModel, action => {
-      if (action.name === 'setAttributeID') {
-        const place: AxisPlace = action.args?.[0],
-          attrID = action.args?.[1]
-        enableAnimation.current = true
-        graphController.handleAttributeAssignment(place, attrID)
-        casesRef.current = graphModel.cases
+    // We only need to make the following connection once
+    useEffect(function passDotsRefToController() {
+      graphController.setDotsRef(dotsRef)
+    }, [dotsRef, graphController])
+
+    // MouseOver events, if over an element, brings up hover text
+    function showDataTip(event: MouseEvent, d: any) {
+      const
+        target = select(event.target as SVGSVGElement)
+      if (target.node()?.nodeName === 'circle' && dataset) {
```

Comment:
We'll need to generalize this to more than dots at some point.

Reply:
Indeed!